### PR TITLE
164 parameter passing robot start

### LINKED ISSUE #164: implement parameter passing to start robot
As an enduser I would like to pass parameters to the robot at the point of starting the robot.
depends on #8 

---

### :thought_balloon: &nbsp; Hints & Thoughts

- when clicking on the start bot icon, an interface should be prompted which asks the user to input the required fields

---

## :white_check_mark: &nbsp; Conditions of satisfaction

- [ ] Prompt user with necessary interface in the web app
- [ ] Enqueue new job object in DB after specifying the parameters.

### PR DESCRIPTION
**Pull request by WolfgangDaniel**
closes #164  

## Changes
When the user clicks on the play button in a robot container he/she will now get redirected to a new page (the robot is not executed immediately anymore). On this new page, the robot execution cockpit, the user starts with step1. In step 1 all the parameters that require a user input before the start are shown as input fields to the user. The parameters are grouped by activities. The user can now fill in all the input that the robot requires (e.g. a filename or a path). Then the user can execute the bot by clicking on the execute robot button or the user can just save the changes and go back to the robot overview. 
When the user clicks on the execute robot button the current local ssot gets uploaded to mongoDB with an upsert call and after that a robot execution job gets triggered using socket.io 

*Info:* I have already rebased the DEV branch with all its nw features and resolved the merge conflicts. 

## Screenshots

Example of robot execution cockpit in case the user has to fill in some parameters:
grafik

Example of robot execution cockpit in case all parameters are hardcoded:
grafik

Example of how the robot execution interface looks after the robot execution has started (interface to be implemented in future user stories):
grafik

## Checklist before merge

**Developer's responsibilities**
* [ ] **Assign** one or two developers
* [ ] **Change code** if reviewer(s) has/have requested it
* [ ] **Pull request build** has passed
* [ ] **tested locally** (in at least chrome & firefox if frontend)
* [ ] updated the **documentation**
* [ ] added **tests** where necessary

### CODE REVIEW

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+const { Step } = Steps;
+const { Title } = Typography;
+
+/**
+ * @description Page, where you can interact with a robot and for example enter input
+ * @category Client
+ * @component
+ */
+const RobotInteractionCockpit = (match) => {
```
Comment (SanJSp) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:19]: what is `match `in this context?

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        Array.prototype.forEach.call(parameterObjects, (parameterObject) => {
+          const { activityId } = parameterObject;
+          let activityName = '';
+          let ssot = sessionStorage.getItem('ssotLocal');
+          ssot = JSON.parse(ssot);
+          Array.prototype.forEach.call(ssot.elements, (elem) => {
+            if (elem.id === activityId) {
+              activityName = elem.name;
+            }
+          });
```
Comment (SanJSp) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:43]: Could [find ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find)be a simpler way to do this?

**Review by SanJSp — APPROVED** [review]
Looks fine from my side :) cool ui you built there :) I've got one idea, and I'd like the linter issues resolved as good as possible. I think you can do that on your own, therefore you have my go :)

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        socket.emit('robotExecutionJobs', { robotId, userId });
+      });
+    } else {
+      alert('Your Bot is not fully configured and can not be executed!');
+    }
+  };
+
+  return (
+    <Layout>
+      <HeaderNavbar selectedKey={4} />
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:102]: At least in my testing this was not correctly displayed, as there is no 4th tab in the header navbar. From my point it should either display that additional tab, or only display the logo, so no editing tabs at all, which I would prefer

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+          setParameterList(activityInformationList);
+        }
+      });
+  }, [robotId]);
+
+  /**
+   * @description Equivalent to ComponentDidMount in class based components
+   */
+  useEffect(
+    () => () => {
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:72]: Why does this need to be a function that returns a function that does something?

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    getAllParametersForRobot(robotId)
+      .then((response) => response.json())
+      .then((parameterObjects) => {
+        const activityInformationList = [];
+        Array.prototype.forEach.call(parameterObjects, (parameterObject) => {
+          const { activityId } = parameterObject;
+          let activityName = '';
+          let ssot = sessionStorage.getItem('ssotLocal');
+          ssot = JSON.parse(ssot);
+          Array.prototype.forEach.call(ssot.elements, (elem) => {
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:39]: Exactly why is this needed as prototype? Shouldn't by the async nature of it ssot be resolved at this point?

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            if (elem.id === activityId) {
+              activityName = elem.name;
+            }
+          });
+          const activityParameter = [];
+          Array.prototype.forEach.call(
+            parameterObject.rpaParameters,
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:46]: What happens if there are no such parameters set?

**Review by kej-jay — CHANGES_REQUESTED** [review]
For what it is doing it looks pretty solid. 
But I have a major issue with the values being set in the actual ssot. Just because you hide the value in the sidebar when the checkbox is set, doesn't mean that the value isn#t actually set for all users, which makes the point of having parameters specified at the robot start unnecessary. For example you can still see the value in the RPAf code.
There should either be a separate object for this or maybe the values should be added to the job object to which they are related. Then they could be parsed by the local client for execution.

I know it would probably require major rewriting of code, but I would strongly vote against using this code the way it is right now, setting the values into the ssot.

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 1 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:18]: This is needed for the Routing. When we click on the play button in RobotOverview a new page will open up: RobotInteractionCockpit. In RobotInteractionCockpit we need some of the information that we have in the RobotOverview page (userId & robotId). Match is an object provided by react routing that allows us to access the parameters we pass to the new page. We access these parameters in the following lines...

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 2 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:43]: Good point. I thought about it too. But in the end it wouldn't save any lines of code and I would get in trouble with asynchronicity again. So I would leave it as it is.

File: client/src/components/content/RobotContainer/RobotContainer.jsx [hunk 7]
```diff
@@ -4,9 +4,9 @@ import { Col, Row, Typography } from 'antd';
 import { PlayCircleOutlined, EditOutlined } from '@ant-design/icons';
 import PropTypes from 'prop-types';
 import { Link } from 'react-router-dom';
-import socket from '../../../utils/socket/socketConnections';
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/content/RobotContainer/RobotContainer.jsx:7]: oh my mistake. The socket import should actually be excluded because we don't need the startRobot function here anymore. I accidentally left the unused startRobot Function in RobotContainer. I fixed it all now. ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 3 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:128]: Thanks for the hint. I agree. I changed the header nav bar now so that only the tab for the robot interaction cockpit is displayed when we are on the RobotInteractionCockpit page. ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 6 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:46]: That is not a problem. In this case just nothing would be added to activityParameter so it would be an empty array. In the component RobotInteractionInputSection in line 19 we are checking if the array activityParameter is empty and just render new parameter input fields if it is not empty.

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 5 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:39]: I am not quite sure if I understand what you mean. I want to iterate here over ssot.elements by using the .foreach function. The problem is that the type of ssot.elements (although it is an array) is Object and .foreach would not work. Thats why I use the Array.prototype.foreach construct. [Here](https://stackoverflow.com/questions/43743560/foreach-vs-array-prototype-foreach-call) is a stackoverflow thread (although not 100% our case here) it explains the use case for Array.prototype.foreach quite well

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 4 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:72]: With _getActivityAndParameterInformation_ we are using an asynchronous callback that uses setState functionality (setParameterList). We want to ensure that this setState call is not executed when the component was already unmounted. Thats why we use _useRef_ and save us with isMounted if the component is mounted. When using useEffect without any dependencies like the useEffect here in line 71, we can return a function. This function will be saved by useEffect and executed when the component unmounts. In our case we are setting isMounted.current = true. And here I actually made a mistake. We should of course set isMounted.current = **false**. I just fixed that. ✅ With that we have ensured that setParameter will not be called when the component was unmounted.

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 3 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:130]: update: I also put the robot overview tab back in the header bar so that we can easily navigate back to the overview page

File: server/services/SsotToRobotParsing/retrieveParameters.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+  return parameterObjects;
+};
+
+/**
+ * @description Update Parameter Objects with new parameters
+ * @param {String} parameterObjects The selection of parameter objects this function will have a look at
+ * @param {String} newParameters New parameters in the form {id, value} that the function will use to update the parameter objects
+ */
+const updateParameterObjects = (parameterObjects, newParameters) => {
```
Comment (kej-jay) [inline server/services/SsotToRobotParsing/retrieveParameters.js:51]: Not use, if this couldn't be done in an async method, would again remove array.prototype

File: server/services/SsotToRobotParsing/retrieveParameters.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  Array.prototype.map.call(parameterObjects, (parameterObject) => {
+    if (parameterObject.rpaParameters.length !== 0) {
+      Array.prototype.map.call(
+        parameterObject.rpaParameters,
+        (currentParameter) => {
+          Array.prototype.forEach.call(newParameters, (newParameter) => {
+            if (
+              // eslint-disable-next-line no-underscore-dangle
+              String(newParameter.parameterId) === String(currentParameter._id)
```
Comment (kej-jay) [inline server/services/SsotToRobotParsing/retrieveParameters.js:60]: Why is this string casting needed?

**Review by kej-jay — CHANGES_REQUESTED** [review]
Overall looks pretty good, minor things and a couple questions. Mostly around async functions. 
I would be glad if you you'd adjust the "normal" ssot-to-rpaf-parser and mark elements which require user input at start with the specified syntax.
Also please check the linting issues and fix them.

File: server/socket/socketHelper.test.js [hunk 10]
```diff
@@ -28,12 +34,27 @@ describe('robot code retrieval', () => {
...
-    const robotCode = await socketHelperFunctions.getRobotCode(testRobotId);
-    expect(robotCode).not.toBeUndefined();
-    expect(robotCode).not.toBeNull();
-    expect(robotCode).toMatch('*** Settings ***');
-    expect(robotCode).toMatch('*** Tasks ***');
+    const robotCode = await socketHelperFunctions.getRobotCodeForJob(
+      testRobotId,
+      testJobId
+    );
+    fs.readFile(
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/socket/socketHelper.test.js:43]: good point. I changed it ✅

File: server/controllers/ssotVariableController.js [hunk 11]
```diff
@@ -4,331 +4,365 @@ const ssotModels = require('../models/singleSourceOfTruthModel.js');
...
+  return mongoose
+    .model('parameter')
+    .findOneAndUpdate(
+      {
+        robotId,
+        activityId,
+      },
+      {
+        rpaParameters: updatedParameters,
+        $unset: { outputVariable: 1 },
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotVariableController.js:68]: This is legacy code. I don't know.

File: server/controllers/ssotRetrievalController.js [hunk 12]
```diff
@@ -184,15 +184,11 @@ exports.overwriteRobot = async (req, res) => {
...
-      .findByIdAndUpdate(
-        updatedSsot['_id'],
-        updatedSsot,
-        {
-          new: true,
-          useFindAndModify: false,
-          upsert: true
-        }
-      )
+      .findByIdAndUpdate(updatedSsot['_id'], updatedSsot, {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/controllers/ssotRetrievalController.js:187]: ✅

File: client/src/utils/attributeAndParamUtils.js [hunk 13]
```diff
@@ -444,6 +447,59 @@ const getParameterForRobotFromDB = async (robotId) => {
   return response;
 };
 
+/**
+ * @description Will set the single parameter in local session storage
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/utils/attributeAndParamUtils.js:451]: yes. I updated the description ✅

File: server/services/SsotToRobotParsing/__tests__/SsotToRobotParser.test.js [hunk 14]
```diff
@@ -96,4 +101,31 @@ describe('Ssot Parsing', () => {
...
+    await dbLoader.loadAttributesInDb();
+    await dbLoader.loadParametersInDb();
+    await dbLoader.loadJobInDb();
+
+    const parserResultString = await parser.parseCodeForJob(
+      testRobotId,
+      testJobId
+    );
+    fs.readFile(
+      './utils/TestingUtils/testRobotFile.txt',
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/__tests__/SsotToRobotParser.test.js:116]: ✅

File: server/services/SsotToRobotParsing/generateCodeBase.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+      activityId: { $in: listOfActivityIds },
+    })
+    .exec();
+
+  return attributeObjects;
+};
+
+/**
+ * @description Generates that basic code that every robot has
+ * @param {Object} ssot The SSoT
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/generateCodeBase.js:81]: I just copied the code. But you are right. I changed it to ssot now ✅

File: server/services/SsotToRobotParsing/retrieveParameters.js (hunk 9 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/retrieveParameters.js:60]: Because it wouldn't work without it. One was of type object and the other a string

File: client/src/components/content/RobotInteractionSections/RobotInteractionComponents/RobotInteractionInputParameter.jsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  infoText,
+  updateParameterValue,
+  parameterId,
+}) => (
+  <>
+    <Text strong>{variableName}</Text>
+    <Input
+      placeholder={variableName}
+      defaultValue=''
+      onChange={(e) => updateParameterValue(parameterId, e.target.value)}
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/content/RobotInteractionSections/RobotInteractionComponents/RobotInteractionInputParameter.jsx:26]: ✅

File: client/src/components/content/RobotInteractionSections/RobotInteractionInputSection.jsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ * @category Client
+ * @component
+ */
+const RobotInteractionInputSection = ({
+  parameterList,
+  updateParameterValue,
+}) => (
+  <Space size='small' direction='vertical' style={{ width: '100%' }}>
+    {parameterList.map((activityInformation) => {
+      if (activityInformation.activityParameter.length !== 0) {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/content/RobotInteractionSections/RobotInteractionInputSection.jsx:19]: ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+      .then((response) => response.json())
+      .then((parameterObjects) => {
+        if (parameterObjects.length !== 0) {
+          const activityInformationList = [];
+          Array.prototype.forEach.call(parameterObjects, (parameterObject) => {
+            const { activityId } = parameterObject;
+            let activityName = '';
+            let ssot = sessionStorage.getItem('ssotLocal');
+            ssot = JSON.parse(ssot);
+            Array.prototype.forEach.call(ssot.elements, (elem) => {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:40]: ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+    const currentParameters = [...parameters];
+    let found = false;
+    for (let i = 0; i < currentParameters.length; i++) {
+      if (currentParameters[i].parameterId === parameterId) {
+        found = true;
+        break;
+      }
+    }
+    if (found) {
+      Array.prototype.map.call(currentParameters, (parameter) => {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:116]: ✅

File: server/services/SsotToRobotParsing/retrieveParameters.js (hunk 8 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/retrieveParameters.js:51]: ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx]: ✅

File: server/services/SsotToRobotParsing/retrieveParameters.js (hunk 8 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/retrieveParameters.js:51]: ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+  const { userId } = match.location.state;
+  const isMounted = useRef(true);
+  const [parameterList, setParameterList] = useState([]);
+  const [currentStep, setCurrentStep] = useState(0);
+  const [parameters, setParameters] = useState([]);
+
+  /**
+   * @description For each activity of the current robot get the id, the name and all the parameters that require a user input
+   */
+  const getActivityAndParameterInformation = useCallback(() => {
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:30]: Why is the useCallback needed here?

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+    getAllParametersForRobot(robotId)
+      .then((response) => response.json())
+      .then((parameterObjects) => {
+        if (parameterObjects.length !== 0) {
+          const activityInformationList = [];
+          Array.prototype.forEach.call(parameterObjects, (parameterObject) => {
+            const { activityId } = parameterObject;
+            let activityName = '';
+            let ssot = sessionStorage.getItem('ssotLocal');
```
[the remark this answers is not in the document]
Comment (kej-jay) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:39]: Please address this

File: client/src/utils/attributeAndParamUtils.js [hunk 22]
```diff
@@ -444,6 +447,59 @@ const getParameterForRobotFromDB = async (robotId) => {
...
+    .then((response) => response.json())
+    .then((data) => {
+      initSessionStorage('parameterLocalStorage', JSON.stringify([]));
+      sessionStorage.setItem('parameterLocalStorage', JSON.stringify(data));
+    });
+
+  initSessionStorage('taskToApplicationCache', JSON.stringify({}));
+  initSessionStorage('availableApplications', JSON.stringify([]));
+  let applicationList = sessionStorage.getItem('availableApplications');
+  applicationList = JSON.parse(applicationList);
```
Comment (kej-jay) [inline client/src/utils/attributeAndParamUtils.js:492]: Same as above, why not do it in one line as everywhere else in the code?

File: server/services/SsotToRobotParsing/generateCodeForRpaTasks.js [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+ * @description Checks whether the given element has a successor element
+ * @param {Object} currentElement Element to check
+ * @returns {Boolean} Value specifies if element has a successor element
+ */
+const successorTasksExist = (currentElement) =>
+  currentElement.successorIds !== undefined &&
+  currentElement.successorIds[0] !== '';
+
+/**
+ * @description Receives an array of all elements and generates the .robot code for the elements recursively.
```
Comment (kej-jay) [inline server/services/SsotToRobotParsing/generateCodeForRpaTasks.js:67]: Please check the documentation, this doesn't seem to match

Comment (kej-jay) [reply server/services/SsotToRobotParsing/generateCodeForRpaTasks.js:67]: I guess it's just parameters and attributes missing

File: server/services/SsotToRobotParsing/retrieveParameters.js [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * @description Retrieve all parameters for a specific job
+ * @param {String} jobId The id of the job
+ */
+const getAllParametersForJob = async (jobId) => {
+  const jobParametersObject = await mongoose
+    .model('job')
+    .findById(jobId, { parameters: 1 });
+  const jobParameters = jobParametersObject.parameters;
```
Comment (kej-jay) [inline server/services/SsotToRobotParsing/retrieveParameters.js:80]: Why this assignment and not return the value right away?

File: server/services/SsotToRobotParsing/retrieveParameters.js [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+  return jobParameters;
+};
+
+/**
+ * @description For all activities in the ssot this method will retrieve the associated parameter objects
+ * @param {Object} ssot The ssot for which the parameters should be retrieved
```
Comment (kej-jay) [inline server/services/SsotToRobotParsing/retrieveParameters.js:86]: jobId parameter documentation is missing

File: server/socket/socketHelperFunctions.js [hunk 26]
```diff
@@ -60,6 +87,7 @@ exports.createJob = async (userId, robotId, status, parameters) => {
   } catch (err) {
     if (err) {
       console.error(err);
+      return undefined;
```
Comment (kej-jay) [inline server/socket/socketHelperFunctions.js:90]: Where is this coming from?

**Review by kej-jay — CHANGES_REQUESTED** [review]

File: server/socket/socketHelperFunctions.js (hunk 26 above)
Comment (WolfgangDaniel) [reply server/socket/socketHelperFunctions.js:90]: es lint wants to have a return here

File: server/services/SsotToRobotParsing/retrieveParameters.js (hunk 25 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/retrieveParameters.js:86]: ✅

File: server/services/SsotToRobotParsing/retrieveParameters.js (hunk 24 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/retrieveParameters.js:80]: we can return it right away. I changed it

File: server/services/SsotToRobotParsing/generateCodeForRpaTasks.js (hunk 23 above)
Comment (WolfgangDaniel) [reply server/services/SsotToRobotParsing/generateCodeForRpaTasks.js:67]: ✅

File: client/src/utils/attributeAndParamUtils.js (hunk 22 above)
Comment (WolfgangDaniel) [reply client/src/utils/attributeAndParamUtils.js:492]: I have really just copy pasted that code from the modeler page. But I think it is a good point. I changed it ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+  const [parameters, setParameters] = useState([]);
+
+  /**
+   * @description For each activity of the current robot get the id, the name and all the parameters that require a user input
+   */
+  const getActivityAndParameterInformation = useCallback(() => {
+    getAllParametersForRobot(robotId)
+      .then((response) => response.json())
+      .then((parameterObjects) => {
+        if (parameterObjects.length !== 0) {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx]: ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+  const { userId } = match.location.state;
+  const isMounted = useRef(true);
+  const [parameterList, setParameterList] = useState([]);
+  const [currentStep, setCurrentStep] = useState(0);
+  const [parameters, setParameters] = useState([]);
+
+  /**
+   * @description For each activity of the current robot get the id, the name and all the parameters that require a user input
+   */
+  const getActivityAndParameterInformation = useCallback(() => {
```
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:30]: I removed Array.prototype with the last commit

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:39]: I think there is no particular reason to do it. I changed it to one line ✅

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 20 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:30]: I think this [article](https://medium.com/@infinitypaul/reactjs-useeffect-usecallback-simplified-91e69fb0e7a3) explains it quite well

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx [hunk 29]
```diff
@@ -0,0 +1,125 @@
...
+  /**
+   * @description Sends a job to the server to execute a specfic robot for a specific user
+   */
+  const startRobot = async () => {
+    const robotIsExecutable = await isRobotExecutable(robotId);
+    if (robotIsExecutable) {
+      setCurrentStep(1);
+      startRobotForUser(userId, robotId, parameters);
+    } else {
+      alert('Your Bot is not fully configured and can not be executed!');
```
Comment (kej-jay) [inline client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:53]: Please disable the linter here

**Review by kej-jay — COMMENTED** [review]
Looks good, only the very small thing. Please check, why the checks do not pass, as they do for the other branches. 
Maybe run npm install locally again and check if there are any changes in the package lock, as those could affect the ci

File: client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx (hunk 29 above)
Comment (WolfgangDaniel) [reply client/src/components/pages/RobotInteractionCockpit/RobotInteractionCockpit.jsx:54]: ✅

**Review by kej-jay — APPROVED** [review]
Looks good from my side now. Thanks for incorporating the changes and please be careful with the conflicting files when merging.
Great job, I'm looking forward to have this functionality in the DEV
